### ISSUE TITLE
Add unmerged Terraform PR 26796: Add workspaces support to HTTP backend

### ISSUE THREAD

**Issue by lenaing** [opentffoundation/opentf#317]
Hello,

This PR adds workspaces support to the HTTP Backend and could benefit numerous state hosting tools.

Here is the initially opened issue to discuss the support : https://github.com/hashicorp/terraform/issues/26797
Here is my PR of an implementation : https://github.com/hashicorp/terraform/pull/26796

Kind regards,

**Comment by roni-frantchi** [comment]
Hey @lenaing  thanks for submitting!

To help us maintain a clear separation between opentf and hashicorp's offerings, we're asking that people describe issues that are in other repositories rather than linking those directly.
I've thus scrubbed out any links to said issues/PRs - if there's any context missing in the description that could help others further understand the issue please provide that in the issue description.  

It is also worth mentioning that as an author of a PR to legacy your are permitted under legacy Terraform CLA to resubmit your code elsewhere as well. Of course a prior RFC would be appreciated

Thanks!

**Comment by mousedownmike** [comment]
Here's a stab at the RFC:

The current [http backend][1] does not provide support for [multiple workspaces][2]. OpenTofu should support the following `terraform workspace` operations when communicating with a `http` backend. This will bring the `http` backend in line with other [backends that support multiple workspaces][3].

* `delete` - Delete a workspace
* `list` - List Workspaces
* `new` - Create a new workspace
* `select` -Select a workspace
* `show` - Show the name of the current workspace


[1]: https://opentofu.org/docs/language/settings/backends/http "Backend Type: http"
[2]: https://opentofu.org/docs/language/state/workspaces "OpenTofu workspaces"
[3]: https://opentofu.org/docs/language/state/workspaces#backends-supporting-multiple-workspaces "Backends Supporting Multiple Workspaces"

**Comment by cam72cam** [comment]
@lenaing apologies for the delay, the core team has primarily been focusing on General Availability up to this point.

As this is a change to a well established protocol, we would like to review this in depth before accepting this issue.  Would you be willing to edit your issue description with additional detail to the changes required for OpenTofu, the http backend protocol, and potential issues that authors of http backend servers could run into?  You don't have to exactly follow the template, but it can be found here: https://github.com/opentofu/opentofu/issues/new?assignees=&labels=rfc%2Cpending-decision&projects=&template=rfc.yml

We see the community support of this issue, but any protocol change like this should be thoroughly described and discussed before we can take it on.

Thank you for your patience and enthusiasm!

**Comment by pellepelster** [comment]
I would love to give this a try, but looking at the current code my feeling is that even proposing and RFC will need some prior discussions with someone more knowledgeable in the overall architecture. 

The current implementation of the `http` state backend looks quite generic, providing a lot of possibilities to adjust addresses, paths and HTTP methods in contrast to the (also HTTP like) backend `cloud` that already seems to support workspaces but has a very strict and more defined way to communicate with the backend.

My gut-feeling is that this will it make hard, to implement a workspace-aware version of the backend, that is backwards-compatible with how the backend is currently used in the wild without giving the backend a hint how to handle workspaces, so this will not be a drop-in replacements that just works.

Following the already established pattern for the `http` backend of providing and endpoint for every operations, feels a little weird to me, since we need a lot of endpoints for the different workspace operations:

* delete
* list
* get
* ...

which would lead to something like this

``` {
   address = "http://myrest.api.com/foo"
   lock_address = "http://myrest.api.com/foo"
   unlock_address = "http://myrest.api.com/foo"
   workspace_list_address = "http://myrest.api.com/foo"
   workspace_get_address = "http://myrest.api.com/foo"
   //...
}
```

I am also having a hard time to imagine use-cases where the endpoints are hosted on different addresses, is this really necessary?
My current idea (in pseudecode) on how to configure workspaces endpoints would look roughly like this:

``` {
   base_address = "http://myrest.api.com"
   workspaces {
     list_path = "/workspaces"
     get_path = "/workspaces/<workspace>"
     lock_path = "/workspaces/<workspace>/lock"
     //...
   }
}
```

The issue is, from what I see in the code: this could lead to the situation where someone uses workspaces, without configuring them in the plugin which is kind of impossible to handle.
But forcing a workspace configuration for the `http` backend would break backwards compatibility, maybe someone can chime in here with some info how this has been handled in the past?

**Comment by apparentlymart** [comment]
I'm curious what folks would think about an alternative path where instead of continuing to layer more complexity on the `http` backend -- where as @pellepelster  points out, each new feature tends to make the configuration _even more_ verbose -- we could instead define a more stringent protocol that requires the server to use a specific URL layout and methods, and thus reduce the configuration burden significantly.

We already have such protocols defined for [provider registries](https://opentofu.org/docs/internals/provider-registry-protocol/), [module registries](https://opentofu.org/docs/internals/module-registry-protocol/), and [issuing credentials for the other protocols to use](https://opentofu.org/docs/internals/login-protocol/).

All of these build on [the remote service discovery protocol](https://opentofu.org/docs/internals/remote-service-discovery/), which means that the minimum possible configuration burden is just a hostname that can return a discovery document. Given just `registry.opentofu.org` we can discover [all of the services available on that host](https://registry.opentofu.org/.well-known/terraform.json) and so we can immediately know how to install the provider `registry.opentofu.org/apparentlymart/assume` without the author having to describe in detail which URLs and methods to use to make requests to the registry.

[The `remote` backend](https://opentofu.org/docs/language/settings/backends/remote/) already uses this service discovery protocol to resolve the argument `hostname`. Today (for historical reasons) it only supports a proprietary protocol that's intended for one vendor's API, but the service discovery protocol would allow us to detect whether a given hostname supports that proprietary API or some new (explicitly-documented) API and provide equivalent functionality against both API designs.

If we were to do this, I expect that we would:

* Define a new service identifier, like `workspaces.v1`, to represent the new protocol.
* Fully document the new protocol, including the URL schemes and methods used for operations like enumerating all of the available workspaces, creating a new workspace, deleting a workspace, taking out a lock, retrieving a state snapshot, etc. The URL schemes would all be defined relative to a base URL taken from the service discovery document, just like with the `providers.v1` and `modules.v1` protocols.
* Extend the `remote` backend so that it continues to support the current proprietary API (for backward-compatibility) but prefers to use `workspaces.v1` if the given hostname declares that it supports it.

Someone who wants to implement their own server would then write server software that implements the protocol, and place the service discovery document on a hostname they control. Let's assume it's `tofu.example.com` for the sake of this hypothetical.

Then the root module could contain a relatively-minimal configuration like this:

```hcl
  backend "remote" {
    hostname     = "tofu.example.com"
    organization = "examplecorp"

    workspaces {
      # (whichever workspace-selection arguments the
      # server wants to accept, depending on the design
      # of that particular implementation.)
    }
  }
```

All of the complexity of which URLs and methods to use would be moved into the protocol definition instead of the backend configuration. A service like [GitLab-managed State](https://docs.gitlab.com/ee/user/infrastructure/iac/terraform_state.html) would need far less per-root-module configuration, since all of the details would move into their server implementation.

We _could_ implement something like this in the `http` backend instead -- which I think is what @pellepelster was considering -- but I think the `remote` backend already has a design far closer to what I've described and its preexisting use of our service discovery protocol means that we can potentially make it work with new additional configuration settings whatsoever: just specify a `hostname` that supports the new protocol and it should "just work".

Vendors that already offer private module/provider registries can add the new service to their existing service discovery doc and then the credentials that their users already have configured should "just work" for accessing the new `workspaces.v1` service too, or they can use `tofu login tofu.example.com` to get such credentials set up relatively easily and they'll work for all three services. (Reusing the same credentials across multiple protocols is a major motivation for this hostname-based discovery approach, so that vendors can offer what appears to be a single cohesive product under their one hostname.)

I'm curious to learn whether this resonates with the folks in this issue, or if you have different use-cases in mind for which this would be a poor fit.

---

(The `remote` backend also potentially supports _remote operations_ under the current protocol. We could potentially also define an extension service for that, like `remoteops.v1`, but that's a considerably more complicated proposition since remote operations are more tightly coupled with the Tofu CLI implementation, so I'd suggest that we save that for a later addition and focus only on state storage for the first round.)

**Comment by ucfchandra** [comment]
Is this issue still active? Or will it become a non-issue if/when workspaces are deprecated and removed?

**Comment by cam72cam** [comment]
We are still considering this issue in the context of future work in backends, but in the mean time, my recommendation is to use early evaluation in the backend configuration to stand in for workspace support.
